### ISSUE TITLE
HiResCAM : counting 0-value attributions maps in GAP and no-GAP networks

### ISSUE THREAD

**Issue by vggls** [jacobgil/pytorch-grad-cam#372]
I have trained two variations of an EfficientNet-B2 network on the following dataset.

[COVID-19 Radiography Database](https://www.kaggle.com/datasets/tawsifurrahman/covid19-radiography-database)

- Network 1: One structured as “conv layers - GAP layer - raw class scores - softmax” (i.e. as downloaded from PyTorch torchvision plus softmax)
- Network 2: One structured as “conv layers - Flatten - raw class scores - softmax” (i.e. I have removed the GAP layer and have flattened the output of the last conv layer as suggested in the HiResCAM paper)

The networks are comparable in terms of generalization ability yielding similar classification report on the test set (consists of 10% data per class and 2119 images it total). In both networks I applied HiResCAM on the correctly classified test images and my findings are as follows :

- Network 1: 1997/2119 test accuracy and 8/1997 attribution maps have only 0 values
- Network 2: 1971/2119 test accuracy and 537/1971 attribution maps have only 0 values

In other words, it seems that replacing GAP with Flattening has a great influence on the quality of the produced maps, as Network 2 produces almost 500 more 0-valued maps.

Is there an intuition/explanation on this ?

```
test_dataset = ...  #loaded via custom code

effnetb2_gap = torch.load('./xrays_efficientnet_b2_gap.pt', map_location='cpu')
effnetb2_flatten = torch.load('./xrays_efficientnet_b2_flatten.pt', map_location='cpu')

gap_instance = HiResCAM(model=effnetb2_gap, target_layers=[effnetb2_gap.features[8][-1]], use_cuda=False)
flatten_instance = HiResCAM(model=effnetb2_flatten , target_layers=[effnetb2_flatten .features[8][-1]], use_cuda=False)

a,b = 0,0
c,d = 0,0

for image, label in test_dataset:

    if int(torch.argmax(effnetb2_gap(image.unsqueeze(0))))==label:
        a+=1
        gap_attributions = gap_instance(input_tensor=image.unsqueeze(0))[0,:,:]
        if np.all(gap_attributions==0):
            b+=1
    
    if int(torch.argmax(effnetb2_flatten(image.unsqueeze(0))))==label:
       c+=1
       flatten_attributions = flatten_instance(input_tensor=image.unsqueeze(0))[0,:,:]
       if np.all(flatten_attributions==0):
           d+=1


print('GAP - {}/{}'.format(b,a))        # gives  8/1997
print('Flatten - {}/{}'.format(d,c))    # gives 537/1971
```

Network 1 is as follows : 

![gap](https://user-images.githubusercontent.com/55101427/209835772-72987113-43e0-4f2f-a29c-de6e3625514d.PNG)

Network 2 is as follows : 

![flatten](https://user-images.githubusercontent.com/55101427/209835771-fc8094ff-3824-4f68-9710-7d7f00aa5600.PNG)
